### DISCUSSION TITLE
PhpStorm - Unresolved function or method route()

### DISCUSSION THREAD

**Discussion by Jacobtims** [tighten/ziggy#595] [Q&A]
### Ziggy version

1.0

### Laravel version

9.19

### Description

I've already have this problem for a while. 
When I use `route()` in my Vue.js file, it is underlined. When I hover over it it says: Unresolved function or method route().

![Bug](https://user-images.githubusercontent.com/75219092/201528611-5bf1b449-c4b2-42d8-b443-18141a2c926e.png)

I'm using the `@routes` on my app.blade.php
How can I get rid of this underline?

### Ziggy call and context

```js
When I use PhpStorm.
```

### Ziggy configuration

```js
Object { url: "http://application.test", port: null, defaults: {}, routes: {…} }
​
defaults: Object {  }
​
port: null
​
routes: Object { "ignition.healthCheck": {…}, "ignition.executeSolution": {…}, "ignition.updateConfig": {…}, … }
​
url: "http://application.test"
```

### Route definition

```php
Route::get('/dashboard', function () {
    return Inertia::render('Dashboard');
})->middleware(['auth', 'verified'])->name('dashboard');
```

**Comment by bakerkretzmar** [comment]
I would guess this is because it thinks there is no `route()` function defined and doesn't know that it's defined globally, but I don't use PHPStorm so I'm not sure. Is there any way you can tell it to ignore that one function?

**Comment by warrickbayman** [comment]
Not sure if you still need this, but if you're using TypeScript, you can do something like this in a declaration file:

```typescript
declare function route(name: string, params?: any): string;
```

If you have TypeScript configured, PHPStorm will pick up that declaration. Not sure if there's a better way.

**Reply by new-user-name** [reply]
I fell back to your solution. Don't know how to make route() visible in Laravel Nova user-defined components.

**Comment by CamKem** [comment]
It would be great if the creator of Ziggy (or another member of the community) could make a plugin for PhpStorm that gives declarative support for route() in Phpstorm so the program know that it is using a global route() function. @taylorotwell

**Reply by Jacobtims** [reply]
Maybe an idea for the Laravel idea plugin @adelf ?

**Reply by adelf** [reply]
It seems I managed to make it work(it wasn't easy, believe me :) ). PhpStorm understands the function and Router class it could return. Laravel Idea with this will be released hopefully in the next week.

<img alt="image" src="https://user-images.githubusercontent.com/2818394/224791401-43055ba7-abf4-4295-b4a8-f4ce41663dd1.png">

**Reply by adelf** [reply]
Well, it's fixed only for js sources. Inside Vue templates - still warnings

<img width="377" alt="image" src="https://user-images.githubusercontent.com/2818394/224793323-7c117bff-9117-485a-a948-b46e1ca76645.png">

**Reply by darakanoit** [reply]
i`m waiting functional on vue component :)

**Reply by adelf** [reply]
Everything looks good now. I'll try to release the next version in the next week.
<img width="682" alt="image" src="https://user-images.githubusercontent.com/2818394/226180848-ee22ac37-c5ec-4bd0-8b22-8688188d2ce8.png">

**Reply by Jacobtims** [reply]
Thx! It works fine in the new update.

**Reply by CamKem** [reply]
Thanks heaps for this addition, it is so much cleaner in PhpStorm to have this function being picked up by thee IDE.
So worth the subscription for Laravel Idea! @adelf

**Comment by new-user-name** [comment]
Hi, thanks for your discussion. What have I to do exactly? I installed 1.6, added @routes to blade.
ts compiles OK, but PHPstorm still says "Unresolved function or method route()".

**Reply by CamKem** [reply]
You need to install the Laravel Idea plugin, it's a really great tool I highly recommend it.
It is paid, but totally worth it for how much it helps & also increases productivity.

**Comment by lucasdcrk** [comment]
For anyone reading later on, another solution is to create a `route.js` file in your utilites folder or similar, with the following content:
```js
// utilities/route.js

import Router from 'ziggy-js/src/js/Router.js'

export function route(name, params, absolute, config) {
  const router = new Router(name, params, absolute, config);

  return name ? router.toString() : router;
}
```

Then import the method in your Vue components (auto imports will work), or use global properties

`import { route } from '@/utilities/route.js'`
